proposals/0000-hypercore.md

```diff
@@ -0,0 +1,274 @@
+
+Let's look at an example for a feed containing four values (chunks). Our tree of hashes will look like this:
+
+```
+0
+  1
+2
+    3 (root node)
+4
+  5
+6
+```
+
+We want to receive and verify the data for 0 (chunk0). To accomplish this, we need to receive:
+
+ - Chunk0
+ - 2, the sibling hash
+ - 5, the uncle hash
+ - 3, the signed root hash
+
+We will first verify the signature on 3. Then, we use the received data to recalculate 3:
+
+```
+0 = h(chunk0)
+2 = (hash received)
+1 = h(0 + 2)
+5 = (hash received)
+3 = h(1 + 5)
+```
+
+If our calculated 3 is equal to our received signed 3, then we know the chunk0 we received is valid.
+
+Since we only need uncle hashes to verify the block, the number of hashes we need is at worst `log2(number-of-blocks)` and the roots of the merkle trees which has the same complexity.
+
+Notice that all new signatures verify the entire dataset since they all sign a merkle tree that spans all data. If a signed update ever conflicts against previously-verified trees, suggesting a change in the history of data, the feed is considered "corrupt" and replication will stop. This serves to disincentivize changes to old data and avoids complexity around identifying the canonical history.
+
+
+# Specifications and parameters
+[specifications-and-parameters]: #specifications-and-parameters
+
+TODO: list what kind of hashes and asymmetric keys are used
+
+TODO: list any parameters (entry sizes)
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## The linear history requirement
+[linear-history-requirement]: #linear-history-requirement
+
+Hypercore assumes a linear history. It can not support branches in the history (multiple entries with the same sequence number) and will fail to replicate when a branch occurs. This means that applications must be extremely careful about ensuring correctness. Users can not copy private keys between devices or processes without strict coordination between them, otherwise they will generate branches and "corrupt" the hypercore.
+
+## Private key risk
+[private-key-risk]: #private-key-risk
+
+Hypercore assumes that ownership of the private key is tantamount to authorship. If the private key is leaked, the author will lose control of the hypercore integrity. If the private key is lost, the hypercore will no longer be modifiable.
+
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+The Hypercore log is conceptually similar to Secure Scuttlebutt's log structure; both are designed to provide a single append-only history and to verify the integrity using only a public key identifier. However, Secure Scuttlebutt uses a Linked List structure with content-hash pointers to construct its log while Hypercore uses a Merkle Tree. This decision increases the amount of hashes computed and stored in Hypercore, but it enables more efficient partial replication of the dataset over the network as trees enable faster comparisons of dataset availability and verification of integrity. (Citation needed?)
```

Comment:
Note the amount it is actually only one hash per block of data in hypercore (even numbers are data, odd ones are hashes)

Reply:
Don't you also have to deal with parent & root hashes?

Reply:
still only one hash stored per data entry. if you include the hash of the data itself is 2x data hashes